vms/platformvm/camino_service.go

```diff
@@ -703,48 +680,93 @@ func (s *Service) GetLastAcceptedBlock(r *http.Request, _ *struct{}, reply *api.
 	return nil
 }
 
-func (s *Service) getKeystoreKeys(creds *api.UserPass, from *api.JSONFromAddrs) (*secp256k1fx.Keychain, error) {
-	// Parse the from addresses
-	fromAddrs, err := avax.ParseServiceAddresses(s.addrManager, from.From)
-	if err != nil {
-		return nil, err
-	}
-
+func (s *Service) getKeystoreKeys(creds *api.UserPass, from *api.JSONFromAddrs) ([]*crypto.PrivateKeySECP256K1R, error) {
 	user, err := keystore.NewUserFromKeystore(s.vm.ctx.Keystore, creds.Username, creds.Password)
 	if err != nil {
 		return nil, err
 	}
 	defer user.Close()
 
-	// Get the user's keys
-	privKeys, err := keystore.GetKeychain(user, fromAddrs)
-	if err != nil {
-		return nil, fmt.Errorf("couldn't get addresses controlled by the user: %w", err)
-	}
+	var keys []*crypto.PrivateKeySECP256K1R
+	if len(from.Signer) > 0 {
+		if len(from.From) == 0 {
+			return nil, errNoKeys
+		}
 
-	// Parse the change address.
-	if len(privKeys.Keys) == 0 {
-		return nil, errNoKeys
+		// Get fake keys for from addresses
+		keys, err = s.getFakeKeys(&api.JSONFromAddrs{From: from.From})
+		if err != nil {
+			return nil, err
+		}
+
+		// Parse the signer addresses
+		signerAddrs, err := avax.ParseServiceAddresses(s.addrManager, from.Signer)
+		if err != nil {
+			return nil, err
+		}
+
+		// Get keys for multiSig
+		keyChain, err := keystore.GetKeychain(user, signerAddrs)
+		if err != nil {
+			return nil, fmt.Errorf("couldn't get addresses controlled by the user: %w", err)
+		}
+		if len(keyChain.Keys) == 0 {
+			return nil, errNoKeys
+		}
+
+		keys = append(keys, nil)
+		keys = append(keys, keyChain.Keys...)
+	} else {
+		// Parse the from addresses
+		fromAddrs, err := avax.ParseServiceAddresses(s.addrManager, from.From)
+		if err != nil {
+			return nil, err
+		}
+
+		// Get the user's keys
+		keyChain, err := keystore.GetKeychain(user, fromAddrs)
+		if err != nil {
+			return nil, fmt.Errorf("couldn't get addresses controlled by the user: %w", err)
+		}
+		if len(keyChain.Keys) == 0 {
+			return nil, errNoKeys
+		}
+
+		keys = keyChain.Keys
 	}
 
 	if err = user.Close(); err != nil {
 		return nil, err
 	}
-	return privKeys, nil
+	return keys, nil
 }
 
-func (s *Service) getFakeKeys(args *api.JSONFromAddrs) (*secp256k1fx.Keychain, error) {
+func (s *Service) getFakeKeys(from *api.JSONFromAddrs) ([]*crypto.PrivateKeySECP256K1R, error) {
```

Comment:
I'm missing the purpose why are we faking the keys.
Please consider to add comment to this fn.